spirv_msl.cpp

```diff
@@ -577,61 +577,87 @@ uint32_t CompilerMSL::get_input_buffer_block_var_id(uint32_t msl_buffer)
 	return ib_var_id;
 }
 
-// Sort the members of the struct type by offset, and pad where needed.
-void CompilerMSL::pad_input_buffer_block(uint32_t ib_type_id)
+// Sort the members of the struct type by offset, and pad & pack
+// members where needed to align MSL members with SPIR-V offsets.
+void CompilerMSL::align_struct(SPIRType &ib_type)
 {
-	auto &ib_type = get<SPIRType>(ib_type_id);
+	uint32_t &ib_type_id = ib_type.self;
 
 	// Sort the members of the interface structure by their offset.
+	// They should already be sorted per SPIR-V spec anyway.
 	MemberSorter member_sorter(ib_type, meta[ib_type_id], MemberSorter::Offset);
 	member_sorter.sort();
 
 	uint32_t curr_offset = 0;
-	for (uint32_t mbr_idx = 0; mbr_idx < ib_type.member_types.size(); mbr_idx++)
+	uint32_t mbr_cnt = uint32_t(ib_type.member_types.size());
+	for (uint32_t mbr_idx = 0; mbr_idx < mbr_cnt; mbr_idx++)
 	{
+		// Align current offset to the current member's default alignment.
+		size_t align_mask = get_declared_struct_member_alignment(ib_type, mbr_idx) - 1;
+		curr_offset = uint32_t((curr_offset + align_mask) & ~align_mask);
+
+		// Test member alignment and, if it needs to be adjusted, depending on which direction,
+		// either pad the current member by adding a dummy padding member that will be declared
+		// before the current member, or mark the previous member as packed, so that it will
+		// consume less space, and the current member can move forward.
+		// There is a situation where, in packing the previous member, the alignment of
+		// that member will change, necessitating the addition of a padding member before it.
+		// This is taken care of automatically because this compiler is multipass. The first
+		// pass will mark the packed member, and the second pass will insert a padding member.
+		// If we ever move to a single-pass design, this will break.
 		uint32_t mbr_offset = get_member_decoration(ib_type_id, mbr_idx, DecorationOffset);
-		uint32_t gap = mbr_offset - curr_offset;
+		int32_t gap = mbr_offset - curr_offset;
```

Comment:
Hmmm...[some reading](http://stackoverflow.com/a/7221449/1678054) I had done before implementing this indicates that wrap-around of uint subtraction IS well defined...and would be cast correctly to the corresponding negative int value.

Nevertheless...I've added the int casts anyway...since it is unambiguous and clearer.